### PR TITLE
Remove numba kernels from `find_index_of_val`

### PR DESCRIPTION
**Pull request by brandon-b-miller**
This PR replaces a few kernels in `cudautils` with equivalent logic composed of libcudf operations. This should have the effect of removing numba from our indexing codepath. 

Contributes to #10706

### CODE REVIEW

File: python/cudf/cudf/utils/cudautils.py [hunk 1]
```diff
@@ -79,23 +30,27 @@ def find_index_of_val(arr, val, mask=None, compare="eq"):
...
+            if arr.dtype in ("float32", "float64"):
+                if np.isnan(val):
+                    locations = is_non_nan(arr)
                 else:
-                    gpu_mark_found_int.forall(found.size)(
-                        arr, val, found, arr.size
-                    )
+                    locations = arr != val
+            else:
+                locations = arr != val
```
Comment (galipremsagar) [inline python/cudf/cudf/utils/cudautils.py:47]: Could be simplified as?:
```
            if arr.dtype in ("float32", "float64") and np.isnan(val):
                locations = is_non_nan(arr)
            else:
                locations = arr != val
```

File: python/cudf/cudf/utils/cudautils.py [hunk 2]
```diff
@@ -79,23 +30,27 @@ def find_index_of_val(arr, val, mask=None, compare="eq"):
...
-            elif compare == "lt":
-                gpu_mark_lt.forall(found.size)(arr, val, found, arr.size)
-            else:
-                if arr.dtype in ("float32", "float64"):
-                    gpu_mark_found_float.forall(found.size)(
-                        arr, val, found, arr.size
-                    )
+    arr = cudf.core.column.as_column(arr, nan_as_null=False)
+    target = cudf.core.column.arange(0, arr.size, dtype="int32")
+    if arr.size > 0:
```
Comment (bdice) [inline python/cudf/cudf/utils/cudautils.py:35]: It looks like `found` is never defined if this conditional is not met, which I suspect would yield a `NameError` in the return line.

File: python/cudf/cudf/utils/cudautils.py [hunk 3]
```diff
@@ -16,56 +17,6 @@
...
-    if i < arr.size:
-        if arr[i] < val:
-            out[i] = i
-        else:
-            out[i] = not_found
-
-
 def find_index_of_val(arr, val, mask=None, compare="eq"):
     """
     Returns the indices of the occurrence of *val* in *arr*
```
Comment (bdice) [inline python/cudf/cudf/utils/cudautils.py:22]: A broader question: why is this function designed like this? The return value isn't what I expect for "Returns the indices of the occurrence of *val* in *arr*".

```python
>>> s = cudf.Series([0, 3, 3, 2, 5, 3])
>>> cudf.utils.cudautils.find_index_of_val(s, 3)
<cudf.core.column.numerical.NumericalColumn object at 0x...>
[
  6,
  1,
  2,
  6,
  6,
  5
]
dtype: int32
```

I expected `[1, 2, 5]` to be the result, not `[6, 1, 2, 6, 6, 5]`. Returning a full column that is the length of the input array indicates that we might have room to improve here.

File: python/cudf/cudf/utils/cudautils.py [hunk 4]
```diff
@@ -79,23 +30,27 @@ def find_index_of_val(arr, val, mask=None, compare="eq"):
...
-                gpu_mark_gt.forall(found.size)(arr, val, found, arr.size)
-            elif compare == "lt":
-                gpu_mark_lt.forall(found.size)(arr, val, found, arr.size)
-            else:
-                if arr.dtype in ("float32", "float64"):
-                    gpu_mark_found_float.forall(found.size)(
-                        arr, val, found, arr.size
-                    )
+    arr = cudf.core.column.as_column(arr, nan_as_null=False)
+    target = cudf.core.column.arange(0, arr.size, dtype="int32")
```
Comment (bdice) [inline python/cudf/cudf/utils/cudautils.py:34]: Can we construct `target` later, closer to where it is needed?

File: python/cudf/cudf/utils/cudautils.py (hunk 3 above)
Comment (bdice) [reply python/cudf/cudf/utils/cudautils.py:22]: We definitely need to refactor this more broadly, after removing numba. We only use `find_index_of_val` in `find_first` and `find_last`, which are in turn only used in one place each:
- https://github.com/rapidsai/cudf/blob/b8c0501d184a63c59cd70aca189b4887ba4f13fd/python/cudf/cudf/core/column/numerical.py#L611
- https://github.com/rapidsai/cudf/blob/b8c0501d184a63c59cd70aca189b4887ba4f13fd/python/cudf/cudf/core/column/numerical.py#L624

I think we have taken on much more complexity here than we really need.

**Review by bdice — CHANGES_REQUESTED** [review]

File: python/cudf/cudf/utils/cudautils.py (hunk 1 above)
Comment (bdice) [reply python/cudf/cudf/utils/cudautils.py:47]: Do we even need to check the dtype, or is it enough to check `np.isnan(val)`?

Comment (bdice) [reply python/cudf/cudf/utils/cudautils.py:47]: If so, we could write a one-liner:
```python
locations = is_non_nan(arr) if np.isnan(val) else arr != val
```

**Comment by bdice** [comment]
@brandon-b-miller Thank you SO MUCH for doing this. This is an area of cudf that I've been wanting to improve for ages.

File: python/cudf/cudf/utils/cudautils.py (hunk 1 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/utils/cudautils.py]: This works great!

File: python/cudf/cudf/utils/cudautils.py (hunk 4 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/utils/cudautils.py]: Done!

File: python/cudf/cudf/utils/cudautils.py (hunk 3 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/utils/cudautils.py:22]: I think these are just fairly old functions for which this is one of the last usages. It's usage in the indexing codepath isn't even right, as it's used to produce results that deviate from pandas in some cases. I expect the function will ultimately be removed entirely as part of @wence- 's broader refactor of indexing, making this PR more of a stopgap as part of the wider MVC effort around organizing numba features. If that's the case then we may as well just wait for the bulk of this to be removed.

Comment (bdice) [reply python/cudf/cudf/utils/cudautils.py:22]: Great. Removing these entirely is my hope as well.

**Review by bdice — APPROVED** [review]

**Review by galipremsagar — APPROVED** [review]

File: python/cudf/cudf/utils/cudautils.py (hunk 3 above)
Comment (wence-) [reply python/cudf/cudf/utils/cudautils.py:22]: > I expected [1, 2, 5] to be the result, not [6, 1, 2, 6, 6, 5]. Returning a full column that is the length of the input array indicates that we might have room to improve here.

Hmm, yes, agreed. But I am rewriting anyway as Brandon says.

**Pull request merged**